velox/exec/HashProbe.cpp

```diff
@@ -189,58 +222,278 @@ void HashProbe::initializeFilter(
+
+  maybeSetupSpillInput(
+      hashBuildResult->restoredPartitionId, hashBuildResult->spillPartitionIds);
+
+  if (table_->numDistinct() == 0) {
+    // If the build table is empty, then inner, right and semi joins return
+    // nothing. Given that, we can call 'noMoreInput()' to skip processing the
+    // probe inputs. But if there is spilling triggered at the build side while
+    // building the table, then we still need to process probe inputs to spill
+    // the probe rows if the corresponding partitions have been spilled at the
+    // build side.
+    if (skipProbeOnEmptyBuild()) {
+      if (!needSpillInput()) {
+        noMoreInput();
       }
-    } else if (
-        (isInnerJoin(joinType_) || isLeftSemiJoin(joinType_) ||
-         isRightSemiJoin(joinType_)) &&
-        table_->hashMode() != BaseHashTable::HashMode::kHash) {
-      // Find out whether there are any upstream operators that can accept
-      // dynamic filters on all or a subset of the join keys. Create dynamic
-      // filters to push down.
-      const auto& buildHashers = table_->hashers();
-      auto channels = operatorCtx_->driverCtx()->driver->canPushdownFilters(
-          this, keyChannels_);
-      for (auto i = 0; i < keyChannels_.size(); i++) {
-        if (channels.find(keyChannels_[i]) != channels.end()) {
-          if (auto filter = buildHashers[i]->getFilter(false)) {
-            dynamicFilters_.emplace(keyChannels_[i], std::move(filter));
-          }
+    }
+  } else if (
+      (isInnerJoin(joinType_) || isLeftSemiJoin(joinType_) ||
+       isRightSemiJoin(joinType_)) &&
+      table_->hashMode() != BaseHashTable::HashMode::kHash && !isSpillInput() &&
+      !hasMoreSpillData()) {
+    // Find out whether there are any upstream operators that can accept
+    // dynamic filters on all or a subset of the join keys. Create dynamic
+    // filters to push down.
+    //
+    // NOTE: this optimization is not applied in the following cases: (1) if the
+    // probe input is read from spilled data and there is no upstream operators
+    // involved; (2) if there is spill data to restore, then we can't filter
+    // probe inputs solely based on the current table's join keys.
+    const auto& buildHashers = table_->hashers();
+    auto channels = operatorCtx_->driverCtx()->driver->canPushdownFilters(
+        this, keyChannels_);
+    for (auto i = 0; i < keyChannels_.size(); i++) {
+      if (channels.find(keyChannels_[i]) != channels.end()) {
+        if (auto filter = buildHashers[i]->getFilter(false)) {
+          dynamicFilters_.emplace(keyChannels_[i], std::move(filter));
         }
       }
     }
   }
+}
+
+bool HashProbe::isSpillInput() const {
+  return spillInputReader_ != nullptr;
+}
+
+void HashProbe::prepareForSpillRestore() {
+  checkRunning();
+  VELOX_CHECK(spillEnabled());
+  VELOX_CHECK(hasMoreSpillData());
+
+  // Wait for table build after prepared for spill restore.
+  auto guard = folly::makeGuard([&] { waitForHashTable(); });
```

Comment:
nit: waitForHashTable name suggests that waiting for the hash table happens synchronously; I wonder if there is a better name since waiting happens asynchronously. 

nit: the use of guard is a bit surprising here; would it make sense to move waitForHashTable() to the caller?